feat(ivy): support i18n without closure

ocombe opened the pull request:
So far using runtime i18n with ivy meant that you needed to use Closure and `goog.getMsg` (or a polyfill). This PR changes the compiler to only output closure code when the compiler option `i18nUseClosure` is set to true. This option is set to false by the compiler-cli when the option `i18nInFile` is used, which is what happens when you load a translations file.
This means that if you use the Angular CLI with ivy and load a translations file, you can use i18n and the application will not throw at runtime.
For now it will not translate your application, but at least you can try ivy without having to remove all of your i18n code and configuration.

petebacondarwin reviewed (approved):
LGTM for ngcc

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -338,7 +338,8 @@ export class NgtscProgram implements api.Program {
           this.reflector, evaluator, scopeRegistry, this.isCore, this.resourceManager,
           this.rootDirs, this.options.preserveWhitespaces || false,
-          this.options.i18nUseExternalIds !== false, this.moduleResolver, this.cycleAnalyzer),
+          this.options.i18nUseExternalIds !== false, !this.options.i18nInFile, this.moduleResolver,
```
nit: I'd propose to define a separate const called `i18nUseClosure` and used it here as an arg. That should improve readability and also we can use it as an anchor for additional documentation on why we do it this way (what you explained in PR description):

```
/**
 *  We define `i18nUseClosure` based on the `i18nInFile` config option value. <here is why>
 *  ...
 */
const i18nUseClosure = !this.options.i18nInFile;
```

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -1566,3 +1566,37 @@ function parseNodes(
+ *
+ * @deprecated this method is temporary & should not be used as it will be removed soon
+ */
+export function i18nConfigureLocalize(options: I18nLocalizeOptions = {
```
This function sets translations hash only. Should we may be call it `i18nSetTranslations` and accept translations map as an argument?

ocombe replied:
It sets translations hash only for now, but I think that we'll need to add the locale_id soon as well

AndrewKushnir reviewed (approved):
@ocombe looks good! Just left a couple minor comments. Thank you.

ocombe commented:
Blocked by https://angular-team.atlassian.net/browse/FW-1076 (bazel issue)

gregmagolan commented:
> Blocked by https://angular-team.atlassian.net/browse/FW-1076 (bazel issue)

Alex got https://github.com/angular/angular/pull/21883 merged which should resolve FW-1076. You should be unblocked on master.

ocombe commented:
@gregmagolan just rebased on master and it doesn't change anything for me. It still doesn't take the tsconfig into consideration. Am I supposed to change something else to the bazel build file to make this work?

mhevery commented on packages/compiler/src/render3/view/api.ts:
```diff
@@ -189,6 +189,11 @@ export interface R3ComponentMetadata extends R3DirectiveMetadata {
+  /**
+   * Whether translation statements should use Closure `goog.getMsg` function or not.
+   */
+  i18nUseClosure: boolean;
```
this flag is bothering me. The way I understand it, it controls if the generated code will use `goog.getMsg` or not. The issue is that we plan to allow people to ship code to NPM, which means that the generated code has to be such that it works in both modes, with and without closure.

mhevery commented on packages/core/test/bundling/todo_i18n/tsconfig.json:
```diff
@@ -0,0 +1,36 @@
+    "suppressTsconfigOverrideWarnings": true
+  },
+  "angularCompilerOptions": {
+    "i18nInFile": "translations.xlf",
```
Can we discuss how passing in `.xlf` file into closure will be compatible with our long term goals.

petebacondarwin reviewed (approved):
still LGTM for ngcc (given that it doesn't seem to touch any ngcc files anymore)

mhevery commented on packages/core/test/bundling/hello_world_i18n/index.html:
```diff
@@ -8,6 +8,10 @@
 
+    <script type="text/javascript">
+      // Don't use Closure `goog.getMsg` for translations
+      window.ngI18nClosureMode = false;
```
This should not be needed. Just make sure that `ng_i18n_closure_mode.ts` is imported from `ng_dev_mode` or some other place which will make it ubiquitous.

mhevery commented on packages/core/test/bundling/todo_i18n/index.html:
```diff
@@ -17,6 +17,10 @@
 
+  <script type="text/javascript">
+    // Don't use Closure `goog.getMsg` for translations
+    window.ngI18nClosureMode = false;
```
same here (should not be needed).

mhevery reviewed (approved):
With a minor quibble.

AndrewKushnir commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -189,64 +188,106 @@ describe('i18n support in the view compiler', () => {
          * @meaning meaningA
          */
-        const $MSG_EXTERNAL_idA$ = goog.getMsg("Content A");
+        var $MSG_EXTERNAL_idA$$APP_SPEC_TS_0$;
```
I just tested that with Closure compiler and here are the findings:

- `goog.getMsg` call output should be assigned to separate/unique const (as it was before)
- other var names mush *not* start with `MSG_` prefix, since it's reserved by Closure for `goog.getMsg` output
- multiple declarations of a var with the same `MSG_` prefix are not allowed

Generated output code should look similar to [what we have in our original design doc](https://github.com/angular/angular/blob/master/packages/core/src/render3/i18n.md#translation-message-retrieval) (with a small correction to avoid `MSG_`-prefixed consts/vars in places other than `goog.getMsg`s invocation):

```
var I18N_EXTERNAL_idA$$APP_SPEC_TS_0;
if (ngI18nClosureMode) {
  const MSG_EXTERNAL_idA$$APP_SPEC_TS_0  = goog.getMsg('Hello World!');
  I18N_EXTERNAL_idA$$APP_SPEC_TS_0 = MSG_EXTERNAL_idA$$APP_SPEC_TS_0;
} else {
  I18N_EXTERNAL_idA$$APP_SPEC_TS_0 = i0.ɵi18nLocalize('Hello World!');
}
```

ocombe replied:
nooo :( I spent so much time updating your tests
well, back to square one then

ocombe replied:
Actually do we need complex variable names like this for anything else than closure? Could we just have:
```
var _c0;
if (ngI18nClosureMode) {
  const MSG_EXTERNAL_idA$$APP_SPEC_TS_0  = goog.getMsg('Hello World!');
  _c0 = MSG_EXTERNAL_idA$$APP_SPEC_TS_0;
} else {
  _c0 = i0.ɵi18nLocalize('Hello World!');
}
```

AndrewKushnir commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -189,64 +188,106 @@ describe('i18n support in the view compiler', () => {
+        else {
+            $MSG_EXTERNAL_idB$$APP_SPEC_TS_1$ = $r3$.ɵi18nLocalize("Title B");
+        }
+        const _c2 = ["title", $MSG_EXTERNAL_idB$$APP_SPEC_TS_1$];
```
It'd be great to keep $$ around consts to test output less fragile (otherwise we'll start checking for specific chars).

AndrewKushnir commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -315,21 +362,35 @@ describe('i18n support in the view compiler', () => {
-        });
+        var $MSG_EXTERNAL_7432761130955693041$$APP_SPEC_TS_3$;
+        if (ngI18nClosureMode) {
+            $MSG_EXTERNAL_7432761130955693041$$APP_SPEC_TS_3$ = goog.getMsg("{$interpolation}", { "interpolation": "\uFFFD0\uFFFD" });
```
Could you plz apply the formatting of the 2nd argument's fields? That makes the output more readable (in case it should be manually verified) and the logic that verifies the output is tolerant to whitespaces.

AndrewKushnir commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -1472,6 +1472,13 @@ describe('ngtsc behavioral tests', () => {
     `);
     env.driveMain();
     const jsContents = env.getContents('test.js');
+    expect(jsContents).toContain(`var MSG_EXTERNAL_8321000940098097247$$TEST_TS_0;
```
I think this check is not needed here. The main idea of the test is to verify the format of const names, not the entire translation block structure. (similar comment to the change below in this file)

AndrewKushnir commented on packages/core/src/util/ng_i18n_closure_mode.ts:
```diff
@@ -0,0 +1,38 @@
+
+declare global {
+  const ngI18nClosureMode: boolean;
+  const goog: {getMsg: (input: string, placeholders?: {[key: string]: string}) => string;};
```
Closure compiler conflicts with this `goog` definition, we can just remove it and use `window.goog` in the checks below if needed.

AndrewKushnir commented on packages/core/src/util/ng_i18n_closure_mode.ts:
```diff
@@ -0,0 +1,38 @@
+export function defineNgI18nClosureMode() {
+  // NOTE: Under Ivy we may have both window & global defined in the Node
+  //    environment since ensureDocument() in render3.ts sets global.window.
+  if (typeof window != 'undefined') {
```
It looks like we already have a function to get the `global this` reference in [packages/core/src/util/global.ts](https://github.com/angular/angular/blob/master/packages/core/src/util/global.ts#L18), so we can just reuse it (and probably update ng_dev_mode.ts logic to make use of it as well).

AndrewKushnir commented on packages/core/src/util/ng_i18n_closure_mode.ts:
```diff
@@ -0,0 +1,15 @@
+ */
+
+import {global} from './global';
+
```
I just tested the latest changes in g3 and here are the results:

1. Closure compiler throws an error at `ngI18nClosureMode`:
```
app.closure.js:16: ERROR - variable ngI18nClosureMode is undeclared
if (ngI18nClosureMode) {
    ^^^^^^^^^^^^^^^^^
```

in order to avoid that, the proposal is to add the following code to the `packages/core/src/util/ng_i18n_closure_mode.ts` script right after the import statement:

```
declare global {
  const ngI18nClosureMode: boolean;
}
```

2. Closure also produces a warning like this:

```
app.closure.js:125: WARNING - constant module$contents$google3$javascript$angular2$tests$i18n_test$app$app_I18N_29 assigned a value more than once.
Original definition at XXX/app.closure.js:119
    I18N_29 = i0.?i18nLocalize("Hello {$interpolation}!", { "interpolation": "\uFFFD0\uFFFD" });
```

it looks like the problem was caused by the misplaced comments/annotations:

```
/**
 * @desc d
 * @meaning m
 */ 
var I18N_29;
if (ngI18nClosureMode) {
    const MSG_EXTERNAL_8071726987696286356$$JAVASCRIPT_ANGULAR2_TESTS_I18N_TEST_APP_APP_TS_30 = goog.getMsg("Hello {$interpolation}!", { "interpolation": "\uFFFD0\uFFFD" });
    I18N_29 = MSG_EXTERNAL_8071726987696286356$$JAVASCRIPT_ANGULAR2_TESTS_I18N_TEST_APP_APP_TS_30;
}
```

Note that the @desc and @meaning tags are not attached to the `const MSG_XXX = goog.getMsg(...)` statement.

AndrewKushnir commented:
@ocombe thanks for additional changes! I did another round of testing in g3:

- demo app (with i18n) was built successfully, generated code looks good
- I was *unable* to fully test the app, because of the issue I described in FW-1140: i18n runtime seems to generate incorrect HTML structure (missing content) in case there are few i18n sections present in a template, like:

```
  <div i18n>Section 1</div>
  <div i18n>Section 2</div>
  <div i18n>Section 3</div>
```

the output is `<div></div><div></div><div>Section 3</div>`.

I verified that this problem exists in current master and understand that this problem is not related to this PR, but it makes impossible to do full round of testing in g3 (to make sure there are no other hidden issues while generating code using Closure Compiler and running it afterwards).

Could you please let me know if this is something you can look at?

cexbrayat commented:
@ocombe Is this still blocked? That would be awesome to get it merged because it is currently blocking external testing :)

cexbrayat commented:
(damn, you pushed when I commented 👌 )

ocombe commented:
it isn't blocked, I just asked @AndrewKushnir to test it against g3 again, hopefully we should be able to merge it now

AndrewKushnir commented:
@ocombe I will test it in g3 asap. I also started a [presubmit](http://test/OCL:236737578:BASE:241753940:1554311423906:fc81342e). Thank you.

AndrewKushnir commented:
Presubmit is successful, basic checks with demo app also look good, but we need to look into the `ngI18nClosureMode` flag in Closure configuration to make sure unnecessary code is tree-shaken away properly.

IgorMinar commented on integration/cli-hello-world-ivy-compat/package.json:
```diff
@@ -28,8 +28,8 @@
-    "@angular-devkit/build-angular": "0.13.5",
-    "@angular/cli": "file:../../node_modules/@angular/cli",
+    "@angular-devkit/build-angular": "^0.800.0-beta.11",
+    "@angular/cli": "^8.0.0-beta.11",
```
can you please not change this. if you need a newer version, then you should update the root package.json and not this one.

ocombe replied:
we talked about this at the framework meeting yesterday and it was decided to only change this in the ivy integration apps for now

IgorMinar replied:
can you please share why?

my concern is that we are making our integration testing infrastructure more fragile if we make this change and we are increasing technical debt. I'd prefer not to make a change like that unless we really really have to (and then there should be a follow up work to remove this temporary hack which should be tracked and scheduled so that we don't forget).

IgorMinar commented on integration/cli-hello-world-ivy-compat/yarn.lock:
```diff
@@ -2,153 +2,141 @@
-    inquirer "6.2.1"
-    opn "5.3.0"
-    semver "5.5.1"
+"@angular/cli@^8.0.0-beta.11":
```
please revert

IgorMinar commented on integration/cli-hello-world-ivy-minimal/package.json:
```diff
@@ -28,8 +28,8 @@
-    "@angular-devkit/build-angular": "0.13.5",
-    "@angular/cli": "file:../../node_modules/@angular/cli",
+    "@angular-devkit/build-angular": "^0.800.0-beta.11",
+    "@angular/cli": "^8.0.0-beta.11",
```
revert

IgorMinar commented on integration/cli-hello-world-ivy-minimal/yarn.lock:
```diff
@@ -2,153 +2,141 @@
-    inquirer "6.2.1"
-    opn "5.3.0"
-    semver "5.5.1"
+"@angular/cli@^8.0.0-beta.11":
```
revert

IgorMinar commented on packages/core/test/bundling/todo_i18n/index.ts:
```diff
@@ -7,10 +7,10 @@
  */
 
 import '@angular/core/test/bundling/util/src/reflect_metadata';
+// TODO(ocombe): replace this with the real runtime i18n service configuration
```
why will this happen? if this really is a TODO then we should likely have an item in jira for this. If this is just a note then make it a note and not a TODO

ocombe replied:
Because `ɵi18nConfigureLocalize` is just temporary (as explained in its api docs) until we can work on the service. We needed something for our tests to make i18n work without closure so that people can test ivy even if they use i18n.

IgorMinar replied:
let me try to explain my question better: if I'm reading this todo in the codebase a month from now, what should I do, how should I interpret it, or how can I learn more about what this comment stands for?

In its current form, all I can do is ask you personally what this is about and where I can learn more, and that doesn't scale.

ocombe replied:
I've updated the todo, let me know if it is still unclear. If not then please approve the PR and run g3sync.

IgorMinar commented on integration/_payload-limits.json:
```diff
@@ -12,7 +12,7 @@
       "uncompressed": {
         "runtime": 1440,
-        "main": 14106,
+        "main": 14261,
```
what is causing this regression? do you know?

ocombe replied:
I think this is the content of `ng_i18n_closure_mode.ts` where we define a global value for the i18n closure mode

IgorMinar reviewed (changes requested):
can you please address the comments I posted and split this PR into two - one for cli update to beta.11 and one for the remaining changes. that way we can isolate the payload size regression and attribute it to this change or an issue in cli. thanks

ocombe commented:
@IgorMinar the size regression was present before the cli change. I tried to update the cli version to see if my change to tree shake `ngI18nClosureMode` could fix the regression (https://github.com/angular/angular-cli/pull/13905) but it didn't change anything (in fact it added 2 bytes).

ocombe commented:
I removed the cli changes, and rebased onto master and it fixed the size regression issue. @IgorMinar can you approve the PR now? :)

IgorMinar reviewed (approved):
lgtm except for that one todo

kara commented:
[presubmit](http://test/OCL:242713975:BASE:242714971:1554837219843:3c6df3f6)

ocombe committed:
feat(ivy): support i18n without closure

So far using runtime i18n with ivy meant that you needed to use Closure and `goog.getMsg` (or a polyfill). This PR changes the compiler to output both closure & non-closure code, while the unused option will be tree-shaken by minifiers.
This means that if you use the Angular CLI with ivy and load a translations file, you can use i18n and the application will not throw at runtime.
For now it will not translate your application, but at least you can try ivy without having to remove all of your i18n code and configuration.

IgorMinar commented:
previews presubmit failed, so here is a new one after the last set of changes: http://test/OCL:243069482:BASE:243069609:1554994952547:fae7a7a0
